python/GafferUI/MultiLineTextWidget.py

```diff
@@ -319,9 +328,57 @@ def __drop( self, widget, event ) :
 class _PlainTextEdit( QtGui.QPlainTextEdit ) :
 
 	def __init__( self, parent = None ) :
-
 		QtGui.QPlainTextEdit.__init__( self, parent )
-
+		self.__fixedLines = None
+		self.__widgetFullyBuilt = False
+		
+	def setFixedLines( self, fixedLines ) :
+		
+		self.__fixedLines = fixedLines
+		
+		self.__computeHeight()
+
+	def getFixedLines( self ) :
+
+		return self.__fixedLines
+
+	def __computeHeight( self ) :
+		
+		# since the height can be based on the font metrics when displaying a fixed number of lines, it can be only
+		# computed properly when the stylesheet has been applied to the widget which happens when the widget 
+		# is about to be shown otherwise there is no need to compute it before that (since that font metrics
+		# would be wrong anyway)
+		if not self.__widgetFullyBuilt :
+			return
+		
+		fixedLines = self.getFixedLines()
+		
+		# when the multiline displays fixed lines
+		if fixedLines is not None :
+
+			# computing the font metrics based on the number of lines
+			height = self.fontMetrics().boundingRect( "M" ).height() * fixedLines
+			
+			# also, we need to compute the widget margins to frame the fixed lines nicely 
+			margin = self.contentsMargins().top() + self.contentsMargins().bottom() + self.document().documentMargin()
+			
+			height += margin
+			
+			self.setFixedHeight(height)
+			
+		# otherwise fallbacks to the default behavior which is auto-expanded height
+		else:
+			# resetting the height
+			QWIDGETSIZE_MAX = 16777215 # qt #define not exposed by PyQt or PySide
+			self.setFixedHeight(QWIDGETSIZE_MAX)
+		
+	def showEvent ( self, event ):
+
+		self.__widgetFullyBuilt = True		
+		self.__computeHeight()
```

Comment:
Ah yes, the joys of taking a stroll through the Qt source code :) It definitely seems necessary to read the source to have any chance of understanding what is going on half the time. On that subject, I just took a look at `QWidget::setSizePolicy()` and it already includes a call to `updateGeometry()`, so I'm not sure if the problem you mentioned in `TextWidget` will need some other fix. Perhaps you could start by making a unit test showing the problem? I've certainly no objection to fixing it, but it seems like we might need to understand it a little more first...